### PR TITLE
Fix updating GitHub actions to major tags that are branches

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
If users have something like in the workflow file

```yaml
    - name: Install vcpkg
      uses: lukka/run-vcpkg@v7
```

And the given action, in this case [`lukka/run-vcpkg`](https://github.com/lukka/run-vcpkg) keeps the major version reference as a branch (as opposed to a tag), then Dependabot would not be able to create updates for that.

This PR fixes the problem.

Closes https://github.com/dependabot/dependabot-core/issues/5017.

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
This one is ready now.

It fixes a pretty common case Dependabot could not yet update! 🎉

File: common/lib/dependabot/git_commit_checker.rb [hunk 1]
```diff
@@ -52,17 +52,17 @@ def pinned?
...
 
-      # If the specified `ref` is actually a branch, we're NOT pinned
-      return false if local_upload_pack.match?(%r{ refs/heads/#{ref}$})
+      # Assume we're pinned unless the specified `ref` is actually a branch
+      return true unless local_upload_pack.match?(%r{ refs/heads/#{ref}$})
 
-      # Otherwise, assume we're pinned
-      true
+      # If the specified `ref` is actually a branch, we're pinned if the branch looks like a version
+      version_tag?(ref)
```
Comment (brrygrdn) [inline common/lib/dependabot/git_commit_checker.rb:59]: ❓ Could there be any unintended consequences of making this change in the `common/lib`?

The release branch pattern is very common to `github-actions` and this makes total sense for that ecosystem, but I'm a little concerned that this change in behaviour may mean that in other languages where people are tracking a version branch for other reasons we will start suggesting upgrades where we haven't before 🤔

Comment (deivid-rodriguez) [reply common/lib/dependabot/git_commit_checker.rb:59]: 🤔 good point!

Seems a very edge case in other ecosystems, and I'm not sure what user's expectation is, but probably they don't want an update? Not really sure...

I made a similar change about a month ago: https://github.com/dependabot/dependabot-core/commit/e1a8c49f24e41b4ad368aff3a7d35b76aa8c96cb, by the way 😬.

What should we do? Should I override `pinned?` with this logic just for Actions?

Comment (deivid-rodriguez) [reply common/lib/dependabot/git_commit_checker.rb:59]: I believe https://github.com/dependabot/dependabot-core/commit/e1a8c49f24e41b4ad368aff3a7d35b76aa8c96cb probably makes sense everywhere (and no issues reported so far, which is good).

Something else I'm observing is that this method returns `true` if an explicit branch argument is given. So I believe this should not affect Bundler, which has an explicit `:branch` option when it comes to dealing with `git` sources. In other words, I think this should only potentially affect ecosystems with an "ambiguous notation" for git refs.

Comment (deivid-rodriguez) [reply common/lib/dependabot/git_commit_checker.rb]: Hei @brrygrdn! Does it look ok now? I'd like to move this one forward if so.

File: common/lib/dependabot/git_commit_checker.rb [hunk 2]
```diff
@@ -52,17 +54,17 @@ def pinned?
...
       return true if local_upload_pack.match?(%r{ refs/tags/#{ref}$})
 
-      # If the specified `ref` is actually a branch, we're NOT pinned
-      return false if local_upload_pack.match?(%r{ refs/heads/#{ref}$})
+      # Assume we're pinned unless the specified `ref` is actually a branch
+      return true unless local_upload_pack.match?(%r{ refs/heads/#{ref}$})
 
-      # Otherwise, assume we're pinned
-      true
+      # TODO: Research whether considering branches that look like versions pinned makes sense for all ecosystems
```
Comment (brrygrdn) [inline common/lib/dependabot/git_commit_checker.rb:60]: 👍🏻

File: common/lib/dependabot/git_commit_checker.rb [hunk 3]
```diff
@@ -23,13 +23,15 @@ class GitCommitChecker
...
-                   requirement_class: nil, version_class: nil)
+                   requirement_class: nil, version_class: nil,
+                   consider_version_branches_pinned: false)
       @dependency = dependency
       @credentials = credentials
       @ignored_versions = ignored_versions
       @raise_on_ignored = raise_on_ignored
       @requirement_class = requirement_class
       @version_class = version_class
+      @consider_version_branches_pinned = consider_version_branches_pinned
```
Comment (brrygrdn) [inline common/lib/dependabot/git_commit_checker.rb:34]: I like the approach of parameterising this rather than overriding it 🥇

**Review by brrygrdn — APPROVED** [review]
Thanks for the update, this looks great! I like that we can now progressively roll out the behaviour to ecosystems where we determine it makes sense rather than having to figure it all out centrally 😄

**Comment by deivid-rodriguez** [comment]
Awesome, thanks so much @brrygrdn!

**Pull request merged**